### PR TITLE
Made function api_search

### PR DESCRIPTION
**Pull request by MrRaccoonUA**
# **What does this pull request do?**

The function is created that takes UPCs and returns: product name, UPCs, and image links

### CODE REVIEW

**Review by aartiukh — CHANGES_REQUESTED** [review]
Please add unit tests

File: main.py [hunk 1]
```diff
@@ -1,10 +1,50 @@
...
+        'Content-Type': 'application/json',
+        'Accept': 'application/json',
+        'user_key': 'only_for_dev_or_pro',
+        'key_type': '3scale'
+    }
+    for code in df[upc_column]:
+        start_time = datetime.now()
+        resp = requests.get('https://api.upcitemdb.com/prod/trial/lookup?upc=' + code, headers=headers)
+        data = json.loads(resp.text)
+        time.sleep(10.5)
```
Comment (aartiukh) [inline main.py:24]: Magic number, make it configurable from function params or create a constant if it doesn't make sense,

File: main.py [hunk 2]
```diff
@@ -1,10 +1,50 @@
...
+import json
+import time
+from datetime import datetime
 
 
 def load_data(csv_file_path):
     return pd.read_csv(csv_file_path)
 
 
+def api_search(df, upc_column):
```
Comment (aartiukh) [inline main.py:12]: Add documentation

File: main.py [hunk 3]
```diff
@@ -1,10 +1,50 @@
...
+    for code in df[upc_column]:
+        start_time = datetime.now()
+        resp = requests.get('https://api.upcitemdb.com/prod/trial/lookup?upc=' + code, headers=headers)
+        data = json.loads(resp.text)
+        time.sleep(10.5)
+        if data['code'] == 'OK':
+            for item in data['items']:
+                df_report = df_report.append({'Code': data['code'], 'UPC': item['ean'], 'Product Name': item['title'],
+                                              'Images links': item['images']}, ignore_index=True)
+                print(df[df[upc_column] == code].index.values, 'Success', 'Iteration time:',
```
Comment (aartiukh) [inline main.py:29]: Use `logging` module instead of `print`

File: main.py [hunk 4]
```diff
@@ -1,10 +1,50 @@
...
+        'key_type': '3scale'
+    }
+    for code in df[upc_column]:
+        start_time = datetime.now()
+        resp = requests.get('https://api.upcitemdb.com/prod/trial/lookup?upc=' + code, headers=headers)
+        data = json.loads(resp.text)
+        time.sleep(10.5)
+        if data['code'] == 'OK':
+            for item in data['items']:
+                df_report = df_report.append({'Code': data['code'], 'UPC': item['ean'], 'Product Name': item['title'],
```
Comment (aartiukh) [inline main.py:27]: Why do we need to call `append` twice? You can put empty image url in case of failure

File: main.py [hunk 5]
```diff
@@ -1,10 +1,50 @@
...
+                                              'Images links': item['images']}, ignore_index=True)
+                print(df[df[upc_column] == code].index.values, 'Success', 'Iteration time:',
+                      datetime.now() - start_time)
+        else:
+            df_report = df_report.append({'Code': data['code'], 'UPC': code}, ignore_index=True)
+            print(df[df[upc_column] == code].index.values, 'Error:', data['code'], 'Iteration time:',
+                  datetime.now() - start_time)
+        df_report['UPC'] = df_report['UPC'].apply(int)
```
Comment (aartiukh) [inline main.py:35]: Please elaborate, why do we need this?

File: main.py [hunk 6]
```diff
@@ -1,10 +1,50 @@
...
+        df_report.to_csv('Report.csv')
```
Comment (aartiukh) [inline main.py:36]: This should be done outside this method, just return dataframe

File: main.py [hunk 7]
```diff
@@ -1,10 +1,50 @@
...
+        df_report.to_json('Report.json', orient="records")
+    return df_report
+
+
+def main_function(csv_file_path, upc_column):
+    df = load_data(csv_file_path)
+    df[upc_column] = df[upc_column].apply(str)
```
Comment (aartiukh) [inline main.py:43]: We should not do this here, move this code to `__main__`

File: test-api-search.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+from pandas._testing import assert_frame_equal
+
+
+class TestStringMethods(unittest.TestCase):
+
+    def test_api_search(self):
+        upc_column_name = 'itemCode'
+        df = pd.read_csv('test-data/test-data.csv')
+        df[upc_column_name] = df[upc_column_name].apply(str)
+        expect_df = pd.read_json('Report.json')
```
Comment (aartiukh) [inline test-api-search.py:13]:
```suggestion
        expected_df_report = pd.read_json('Report.json')
```

File: test-api-search.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import pandas as pd
+from pandas._testing import assert_frame_equal
+
+
+class TestStringMethods(unittest.TestCase):
+
+    def test_api_search(self):
+        upc_column_name = 'itemCode'
+        df = pd.read_csv('test-data/test-data.csv')
+        df[upc_column_name] = df[upc_column_name].apply(str)
```
Comment (aartiukh) [inline test-api-search.py:12]: Why do we need this?

**Review by aartiukh — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed

File: main.py (hunk 3 above)
Comment (MrRaccoonUA) [reply main.py]: Done

File: main.py (hunk 5 above)
Comment (MrRaccoonUA) [reply main.py:38]: Fixed

File: main.py (hunk 6 above)
Comment (MrRaccoonUA) [reply main.py]: Done

File: main.py (hunk 7 above)
Comment (MrRaccoonUA) [reply main.py]: Done

File: test-api-search.py (hunk 8 above)
Comment (MrRaccoonUA) [reply test-api-search.py]: Done

File: test-api-search.py (hunk 9 above)
Comment (MrRaccoonUA) [reply test-api-search.py:12]: This is the input to the unittest function

File: main.py (hunk 2 above)
Comment (MrRaccoonUA) [reply main.py:18]: Done

File: main.py [hunk 10]
```diff
@@ -1,10 +1,53 @@
...
+                    format='%(asctime)s.%(msecs)03d[%(name)s:%(lineno)d] %(levelname)s ''%(''message)s',
+                    datefmt='%Y-%m-%d %I:%M:%S')
 
 
 def load_data(csv_file_path):
     return pd.read_csv(csv_file_path)
 
 
+def api_search(df, upc_column):
+    """
```
Comment (aartiukh) [inline main.py:19]: Can we use list here?

File: main.py [hunk 11]
```diff
@@ -1,10 +1,53 @@
...
+    :param df: Dataframe with UPCs that we need to lookup
+    :param upc_column: Name of the column with UPCs in df
+    :return: df_report - this is the answer from API containing requested data
+    """
+    rate_limit = 11
+    df_report = pd.DataFrame(data={'Code': [], 'UPC': [], 'Product Name': [], 'Images links': []})
```
Comment (aartiukh) [inline main.py:25]: Remove code, rename to `df_report` to `df_upc_images`

File: main.py [hunk 12]
```diff
@@ -1,10 +1,53 @@
...
     return pd.read_csv(csv_file_path)
 
 
+def api_search(df, upc_column):
+    """
+    :param df: Dataframe with UPCs that we need to lookup
+    :param upc_column: Name of the column with UPCs in df
+    :return: df_report - this is the answer from API containing requested data
+    """
+    rate_limit = 11
```
Comment (aartiukh) [inline main.py:24]: API_COOL_DOWN_TIMEOUT

File: main.py [hunk 13]
```diff
@@ -1,10 +1,53 @@
...
+        df_report['UPC'] = df_report['UPC'].apply(int)
+    return df_report
+
+
 if __name__ == '__main__':
     csv_file_name = 'test-data/test-data.csv'
-    load_data(csv_file_name)
+    upc_column_name = 'itemCode'
+    df_upc = load_data(csv_file_name)
+    df_upc[upc_column_name] = df_upc[upc_column_name].apply(str)
```
Comment (aartiukh) [inline main.py:51]: Use list here

**Review by aartiukh — CHANGES_REQUESTED** [review]

File: main.py (hunk 10 above)
Comment (MrRaccoonUA) [reply main.py:19]: Yes, we can. Done

File: main.py (hunk 11 above)
Comment (MrRaccoonUA) [reply main.py:25]: Done

File: main.py (hunk 12 above)
Comment (MrRaccoonUA) [reply main.py:24]: Done

File: main.py (hunk 13 above)
Comment (MrRaccoonUA) [reply main.py:51]: Done

File: main.py [hunk 14]
```diff
@@ -1,10 +1,55 @@
...
+        logger.info(f'{list_of_upc.index(code)} Status: {data["code"]}. '
+                    f'Iteration time: {datetime.now() - start_time}')
+        df_upc_images['UPC'] = df_upc_images['UPC'].apply(int)
+    return df_upc_images
+
+
 if __name__ == '__main__':
     csv_file_name = 'test-data/test-data.csv'
-    load_data(csv_file_name)
+    df_upc = load_data(csv_file_name)
```
Comment (aartiukh) [inline main.py:49]: Let's remove this csv file here and allow user to set the UPD list manually or read from `txt` file.

File: main.py [hunk 15]
```diff
@@ -1,10 +1,55 @@
...
     return pd.read_csv(csv_file_path)
 
 
+def api_search(list_of_upc):
+    """
+
+    :param list_of_upc: list of the UPCs that we need to lookup
+    :return:
+    """
+    api_cool_down_timeout = 11
```
Comment (aartiukh) [inline main.py:24]: UPPER_CASE pls for constants

**Review by aartiukh — CHANGES_REQUESTED** [review]

File: main.py (hunk 15 above)
Comment (MrRaccoonUA) [reply main.py]: Fixed

File: main.py (hunk 14 above)
Comment (MrRaccoonUA) [reply main.py]: Done

**Review by aartiukh — APPROVED** [review]

**Pull request merged**